RV-Monitor/Kernel_Safety_Claims_by_Runtime_Verification_Monitors.md

```diff
@@ -0,0 +1,542 @@
+
+
+
+* The hierarchical safety analysis of the Kernel and associated safety requirements definitions can be limited to code functionally contributing to the allocated safety reqs (class A above).  
+* The Runtime Verification Monitor provides an architectural protection mechanisms against Kernel interferences following a safety analysis of the resources to be monitored against interference failure modes  
+
+
+### When is it valuable to use RV Monitors?
+
+From a conceptual perspective the RV Monitors introduce redundancy inside the Kernel, it is therefore valuable to use them as qualification measure when the following criteria are met:
+
+
+
+* The complexity of the Automata model is significantly lower than the code being monitored (i.e. when the code is capable of providing many functionalities under many different condition and states but only a specific subset of its potentials are used in a safety context); this implies that from a FuSa qualification effort perspective and continuous certification perspective dealing with a significantly smaller code baseline is cheaper and faster and hence it is better to make a systematic capability claim on the RV Monitor code instead of spending effort on proving the same for the monitored code.
+* The expected behavior of the monitored code is comprehensively analyzed in how it would meet the allocated safety requirements. This is needed to guarantee that the monitor is monitoring all the critical states and data for safety reasons and to avoid the monitor triggering due to valid states and data not being analyzed. 
+* The specific use case being evaluated is compatible with ex-post-facto treatment of the failure: by design the RVM will be able to detect the failure only after it has taken place and it cannot avoid it (it works as detection measure, not as prevention one) .
+* It is possible to determine that the RVM shall always trigger before the detected failure propagates in a dangerous way to the safety workloads being monitored (avoid failure propagation)
+
+
+# Integration of Runtime Verification Monitors with the Kernel
+
+A monitor is the central part of the runtime verification of a system. The monitor stands in between the formal specification of the desired (or undesired) behavior, and the trace of the actual system.
+
+In Linux terms, the runtime verification monitors are encapsulated inside the RV monitor abstraction. A RV monitor includes a reference model of the system, a set of instances of the monitor defined at development time (e.g. per-cpu monitor, per-task monitor, and so on), and the helper functions that glue the monitor to the system. Depending on both the parameters being monitored and the specific hardware features available, different approaches can provide flexibility in managing the need of monitoring vs the overhead that it might introduce. More on this later.
```

Comment:
If they are supported, I would add also device drivers, since they can be the hardest to qualify, when it comes to detecting interference.

Reply:
Add where? .... RVM can be designed to monitor any custom portion of Kernel code ranging from few files of a driver or subsystem, to an entire driver or subsystem and even to a group of drivers and subsystems (like in the example we made for the WTD...)

Reply:
wher you have the per-cpu, per-task list, it might seem that the RVM covers only system level features, but not peripherals.  For the principle of least surprise (in the reader) I'd rather mention also device drvers.

Reply:
I can rephrase as <<A RV monitor includes a reference model of the system or of a subset of the system (e.g. a model of a specific subsystem or of a device driver), a set of instances of the monitor defined at development time (e.g. per-cpu monitor, per-task monitor, and so on) ... >> ...WDYt ?

Reply:
ok